tests/README.md

```diff
@@ -0,0 +1,124 @@
+
+## VirtualNet
+
+Core of most tests is the `net::VirtualNet` struct, which simulates a network of nodes all running an instance of a distributed algorithm. Messages sent by these nodes are queued by the network and dispatched each time the network is advancing one iteration, commonly referred to as being "cranked". Each time the network is cranked, a buffered message is delivered to its destination node and processed.
+
+Virtual networks can also host an adversary that can affect faulty nodes (which are tracked automatically) or reorder queued messages.
+
+Use the `NetBuilder` to create a new network:
+
+```rust
+// Create a network of 10 nodes, out of which 3 are faulty.
+let mut net = NetBuilder::new(0..10)
+    .num_faulty(3)
+    .using(move |node| { DynamicHoneyBadger::builder().build(node.netinfo) })
+    .build()
+    .expect("could not construct test network");
+```
+
+Algorithms that return a `Step` upon construction should use `using_step` instead.
+
+### Sending input
+
+Send `Input` to any `VirtualNet` node using the `send_input` method:
+
+```rust
+let input = ...;
+let _step = net.send_input(123, input).expect("algorithm failed");
+```
+
+While the resulting step is returned, it needn't be processed to keep the network going, as its messages are automatically added to the queue.
+
+Instead of targeting a node in particular, the same input can be sent to all nodes:
+
+```rust
+net.broadcast_input(input).expect("algorithm failed");
+```
+
+### Cranking the network
+
+The network advances through the `crank()` function, on every call
+
+1. the adversary is given a chance to re-order<sup>1</sup> the message queue,
+1. the next message in the queue is delivered to its destination node (if the node is non-faulty) or the adversary (if the node is faulty),
+1. all messages from the resulting step are queued,
+1. and the resulting step (or error) is returned.
+
+If there were no messages to begin with, `None` is returned instead.
+
+<sup>1</sup>: Due to some implementation deficiencies it is possible for an adversary to mutate any part of VirtualNet (i.e. to change things beyond the scope of our adversary model). While this will be addressed in future versions, it is currently up to the test implementor to ensure that adversaries are not more powerful than they are supposed to be.
+
+Cranking can be done manually:
+
+```rust
+let step = net.crank()
+              .expect("expected at least one messages")
+              .expect("algorithm error");
+
+// Shorthand:
+let step = net.crank_expect();
+```
+
+For convenience, an iterator interface is also available:
+
+```rust
+for res in net {
+    let (node_id, step) = res.expect("algorithm error");
+    // ...
+}
+```
+
+This has the drawback that access to the network is not available between cranks, as it is borrowed inside the for-loop. A common workaround is using a while loop instead:
```

Comment:
"Cranking" is an application of a crank. "Between cranks" doesn't sound right. Better said "between applications of `VirtualNet::crank`".

Reply:
Clarified.